components/merkledb/src/views/metadata.rs

```diff
@@ -212,135 +261,128 @@ impl<T: IndexAccess> IndexesPool<T> {
+    fn set_len(&mut self, len: u64) {
+        self.0.put_or_forget(&(), len);
     }
 
     fn create_index_metadata<V>(
         &mut self,
         index_name: &[u8],
         index_type: IndexType,
-    ) -> IndexMetadata<V>
+    ) -> (IndexMetadata<V>, bool)
     where
-        V: BinaryAttribute + Default + Copy,
+        V: BinaryAttribute,
     {
         let len = self.len();
-
         let metadata = IndexMetadata {
-            index_type,
             identifier: len,
-            state: V::default(),
+            index_type,
+            state: None,
         };
-
-        self.0.put(index_name, metadata.to_bytes());
+        let is_phantom = !self.0.put_or_forget(index_name, metadata.to_bytes());
         self.set_len(len + 1);
-        metadata
+        (metadata, is_phantom)
     }
 }
 
 /// Wrapper struct to manipulate `IndexMetadata` for an index with provided `index_name`.
-/// Metadata value is cached for faster access.
-pub struct IndexState<T, V>
-where
-    V: BinaryAttribute + Default + Copy,
-    T: IndexAccess,
-{
-    index_access: T,
-    index_name: Vec<u8>,
-    cache: Cell<IndexMetadata<V>>,
-    is_new: bool,
+#[derive(Debug)]
+pub struct ViewWithMetadata<T: RawAccess> {
+    view: View<T>,
+    metadata: IndexMetadata,
+    index_full_name: Vec<u8>,
+    is_phantom: bool,
 }
 
-impl<T, V> IndexState<T, V>
+impl<T> ViewWithMetadata<T>
 where
-    V: BinaryAttribute + Default + Copy,
-    T: IndexAccess,
+    T: RawAccess,
 {
-    fn new(index_access: T, index_name: Vec<u8>, metadata: IndexMetadata<V>, is_new: bool) -> Self {
-        Self {
-            index_access,
-            index_name,
-            cache: Cell::new(metadata),
-            is_new,
+    pub(crate) fn get_or_create(
+        index_access: T,
+        index_address: &IndexAddress,
+        index_type: IndexType,
+    ) -> Result<Self, Self> {
+        assert_valid_name(index_address.name());
+        // Actual name.
+        let index_name = index_address.name.clone();
+        // Full name for internal usage.
+        let index_full_name = index_address.fully_qualified_name();
+
+        let mut pool = IndexesPool::new(index_access.clone());
+        let mut is_phantom = false;
+        let metadata = pool.index_metadata(&index_full_name).unwrap_or_else(|| {
+            let (metadata, phantom_flag) = pool.create_index_metadata(&index_full_name, index_type);
+            is_phantom = phantom_flag;
+            metadata
+        });
+        let real_index_type = metadata.index_type;
+        let mut index_address = metadata.index_address();
+        // Set index address name, since metadata itself doesn't know it.
+        index_address.name = index_name;
+        let this = Self {
+            view: View::new(index_access, index_address),
+            metadata,
+            index_full_name,
+            is_phantom,
+        };
+        if real_index_type == index_type {
+            Ok(this)
+        } else {
+            Err(this)
```

Comment:
I guess an explanation (maybe in docstring) what `Err(this)` does mean and why if it's an error we still return the value would be nice.